lucet-wasi/src/main.rs

```diff
@@ -182,22 +183,12 @@ fn run(config: Config) {
         for (dir, guest_path) in config.preopen_dirs {
             ctx = ctx.preopened_dir(dir, guest_path);
         }
-        let mut inst = region
-            .new_instance_builder(module as Arc<dyn Module>)
+        region
+            .new_instance_builder(module as Arc<dyn Module>, config.entrypoint)
             .with_embed_ctx(ctx.build().expect("WASI ctx can be created"))
             .build()
             .expect("instance can be created");
-
-        match inst.run(config.entrypoint, &[]) {
-            // normal termination implies 0 exit code
-            Ok(_) => 0,
-            Err(lucet_runtime::Error::RuntimeTerminated(
-                lucet_runtime::TerminationDetails::Provided(any),
-            )) => *any
-                .downcast_ref::<lucet_wasi::host::__wasi_exitcode_t>()
-                .expect("termination yields an exitcode"),
```

Comment:
This looks like it has the unexpected side-effect of always returning an exit code of 0, even if the wasm that was executed returned a non-0 exit code. If that doesn't fail a test I suspect we might be missing tests :)

Just because you mentioned the `panic!` here: the panic case is for an aytpical runtime exit, such as a timeout or the runtime reaching a known-invalid state, in contrast to an execution where the module ran, everything worked fine, and the program that ran reports an error

Reply:
Thanks for the feedback.  I agree it will always return 0, but in effect the statement before will error out in erroneous cases.   
region
            .new_instance_builder(module as Arc<dyn Module>, config.entrypoint)
            .with_embed_ctx(ctx.build().expect("WASI ctx can be created"))
            .build()
            .expect("instance can be created");
Thus if a valid handle is returned we are exit 0.  At least that was my thinking but perhaps I am misunderstanding more profound.

Reply:
I may be misunderstanding you first comment about _start not being explicitly referenced by the go wasm but I believe it is.  lucetc seems to be acquiring it from its export table and calling what it believes is the correct function index (in my case that was 851). The problem that I see is that the go linker is creating a duplicate record for wasi_fd_write in imports causing the indexing to be off by one and having the lucet-wasi code end up in function 850 causing a barf.  I stepped through wasmtime to see how it was avoiding this issue, and it is doing something similar to the approach I took.  I am not certain lucet-wasi is folding duplicate imports properly. Wasmtime doesn't do it either.  At first I tried to alter the go linker to avoid putting out dupes, but that is tricky and while I could figure out how to remove them from imports table I could not correctly resolve all the old references to the removed duplicate.  I am really a C++ dev pretending to be a Go dev learning Rust hacking on compilers and linkers.  Hope that clarifies.  If you like I can attach sample go code and resultant wasm.